lib/wasi/src/state/mod.rs

```diff
@@ -978,46 +976,23 @@ impl WasiFs {
                 _ => unreachable!("Preopened FD that's not a directory or the root"),
             };
             // stem path based on it
-            if let Ok(rest) = path.strip_prefix(po_path) {
-                // if any path meets this criteria
-                // (verify that all remaining components are not symlinks except for maybe last? (or do the more complex logic of resolving intermediary symlinks))
-                // return preopened dir and the rest of the path
-
-                return Ok((*po_fd, rest.to_owned()));
-            }
-        }
-        Err(__WASI_EINVAL) // this may not make sense
-    }
-
-    // if this is still dead code and the year is 2020 or later, please delete this function
-    #[allow(dead_code)]
-    pub(crate) fn path_relative_to_fd(
-        &self,
-        fd: __wasi_fd_t,
-        inode: Inode,
-    ) -> Result<PathBuf, __wasi_errno_t> {
-        let mut stack = vec![];
-        let base_fd = self.get_fd(fd)?;
-        let base_inode = base_fd.inode;
-        let mut cur_inode = inode;
-
-        while cur_inode != base_inode {
-            stack.push(self.inodes[cur_inode].name.clone());
-            match &self.inodes[cur_inode].kind {
-                Kind::Dir { parent, .. } => {
-                    if let Some(p) = parent {
-                        cur_inode = *p;
-                    }
+            if let Ok(stripped_path) = path.strip_prefix(po_path) {
+                // find the max
+                let new_prefix_len = po_path.as_os_str().len();
+                // we use >= to favor later preopens because we iterate in order
+                // whereas WASI libc iterates in reverse to get this behavior.
+                if new_prefix_len >= max_prefix_len {
+                    max_prefix_len = new_prefix_len;
+                    max_stripped_path = Some(stripped_path);
+                    max_po_fd = Some(*po_fd);
                 }
-                _ => return Err(__WASI_EINVAL),
             }
         }
-
-        let mut out = PathBuf::new();
-        for p in stack.iter().rev() {
-            out.push(p);
+        if max_prefix_len == 0 {
+            Err(__WASI_EINVAL) // this may not make sense depending on where it's called
+        } else {
+            Ok((max_po_fd.unwrap(), max_stripped_path.unwrap().to_owned()))
```

Comment:
(I am very new, but I figure I should introduce myself by bein useful with a code review!)

I wonder if it is possible to structure this so the unwraps are not required. It seems like this function pivots once a candidate is found, would it be possible to fold the three parameters into a single Result value, and have the last step here be over the existing result methods instead of using unwrap?

The unwrap was a red flag for me, and I almost posted about the empty-list case, however these unwraps are safe because the max_prefix_len check and that if it's not null, these two are populated. Recognizing that requires me to mull over the state machine to figure out that the three are correlated, where as I feel the code could directly tell me that.

Either way, still works the same, just a clarity for me!

Reply:
Yeah, this is a good observation! I agree, but it's a bit tricky here.

Sorry by the way -- I missed these comments before -- reviewing them now, I'll make a new PR with any updates I make. Thanks!

Reply:
In theory, we can solve this problem with a typed state machine, I'm just not sure it's really worth it given the size of this function. If it were larger, I'd definitely be in favor though. The thing is it feels like I'll at least double the amount of code to do it that way...

Like

```Rust
enum BaseFdAndRelPath {
  None,
  BestMatch  {
    fd: __wasi_fd_t,
    rel_path: PathBuf,
    max_seen: usize,
  }
}

Impl  BaseFdAndRelPath {
  fn max_seen(&self) -> usize {
     match self {
       Self::None => 0,
       Self::BestMatch { max_seen, .. } => max_seen,
     }
  }
}
```

Actually, that's not too bad with the method... I'll try it out

Reply:
Something like https://doc.rust-lang.org/std/cmp/fn.max_by.html could also work here, unfortunately it's not stable. That would lead to simplified code too.  I'm pretty happy with the result I have, I do think the code is more clear now! Thanks!

The above `Enum` that I posted could definitely be made generic and reused with generic data for finding the most X of a collection of things, could probably even move control flow into a method on this type

Reply:
Pushed it up to #2325

> (I am very new, but I figure I should introduce myself by bein useful with a code review!)

Nice to meet you and thanks for the suggestions! We have a community slack as well where you can ping us if we ever miss something like this. If you're interested in hacking on some Wasm / Wasmer stuff, feel free to reach out!